airflow/models/serialized_dag.py

```diff
@@ -73,8 +73,7 @@ class SerializedDagModel(Base):
 
     dag_runs = relationship(
         DagRun,
-        primaryjoin=dag_id == DagRun.dag_id,
-        foreign_keys=dag_id,
+        primaryjoin=dag_id == foreign(DagRun.dag_id),
```

Comment:
I am still not sure if we explicitly require `foreign` keyword over here.

Can you explain why it is needed?

based on the 2nd section in https://docs.sqlalchemy.org/en/13/orm/join_conditions.html#creating-custom-foreign-conditions

>An alternative syntax to the above is to use the foreign() and remote() annotations, inline within the relationship.primaryjoin expression. This syntax represents the annotations that relationship() normally applies by itself to the join condition given the relationship.foreign_keys and relationship.remote_side arguments. These functions may be more succinct when an explicit join condition is present, and additionally serve to mark exactly the column that is “foreign” or “remote” independent of whether that column is stated multiple times or within complex SQL expressions:

Reply:
You are right. I tested and we don't need that. I am new to Airflow and SQLAlchemy tbh :).

Reply:
Looks like the CI failed with following error:

```
sqlalchemy.exc.InvalidRequestError: One or more mappers failed to initialize - can't proceed with initialization of other mappers. Triggering mapper: 'mapped class SerializedDagModel->serialized_dag'. Original exception was: Could not locate any relevant foreign key columns for primary join condition 'serialized_dag.dag_id = dag_run.dag_id' on relationship SerializedDagModel.dag_runs.  Ensure that referencing columns are associated with a ForeignKey or ForeignKeyConstraint, or are annotated in the join condition with the foreign() annotation.
```

So indeed it looks like we do need `foreign`

Reply:
or maybe it just requires the following:

```
primaryjoin="SerializedDagModel.dag_id == foreign(DagRun.dag_id)",
```

based on example in https://stackoverflow.com/a/57454638/5691525